mpc-core/src/protocols/rep3_ring/gadgets/sort.rs

```diff
@@ -0,0 +1,625 @@
+                rand.push(beta - a);
+                result.push(Rep3RingShare::new_ring(a, RingElement::zero()));
+            }
+            io_context.network.send_many(PartyID::ID1, &rand)?;
+            let rcv: Vec<RingElement<T>> = io_context.network.recv_prev_many()?;
+            for (res, (r1, r2)) in result.iter_mut().zip(rcv.into_iter().zip(rand)) {
+                res.b = r1 + r2;
+            }
+            result
+        }
+    };
+    Ok(result)
+}
+
+fn shuffle_field<F: PrimeField, N: Rep3Network>(
+    pi: &[Rep3RingShare<PermRing>],
+    input: &[Rep3PrimeFieldShare<F>],
+    io_context: &mut IoContext<N>,
+) -> IoResult<Vec<Rep3PrimeFieldShare<F>>> {
+    let len = pi.len();
+    debug_assert_eq!(len, input.len());
+    let result = match io_context.id {
+        rep3::id::PartyID::ID0 => {
+            // has p1, p3
+            let mut alpha_1 = Vec::with_capacity(len);
+            let mut alpha_3 = Vec::with_capacity(len);
+            let mut beta_1 = Vec::with_capacity(len);
+            for a in input {
+                let (alpha_1_, alpha_3_) = io_context.random_fes::<F>();
+                alpha_1.push(alpha_1_);
+                alpha_3.push(alpha_3_);
+                beta_1.push(a.a + a.b);
+            }
+            // first shuffle
+            let mut shuffled_1 = Vec::with_capacity(len);
+            for (pi, alpha) in pi.iter().zip(alpha_1.iter()) {
+                let pi_1 = pi.a.0 as usize;
+                shuffled_1.push(beta_1[pi_1] - alpha);
+            }
+            // second shuffle
+            let mut shuffled_3 = alpha_1;
+            for (des, (pi, alpha)) in shuffled_3.iter_mut().zip(pi.iter().zip(alpha_3)) {
+                let pi_3 = pi.b.0 as usize;
+                *des = shuffled_1[pi_3] - alpha;
+            }
+            io_context.network.send_next_many(&shuffled_3)?;
+
+            // Opt Reshare
+            let mut result = Vec::with_capacity(len);
+            for _ in 0..len {
+                let (a, b) = io_context.random_fes::<F>();
+                result.push(Rep3PrimeFieldShare::new(a, b));
+            }
+            result
+        }
+        rep3::id::PartyID::ID1 => {
+            // has p2, p1
+            let mut alpha_1 = Vec::with_capacity(len);
+            let mut beta_2 = Vec::with_capacity(len);
+            for a in input {
+                let alpha_1_ = io_context.rngs.rand.random_field_element_rng2::<F>();
+                alpha_1.push(alpha_1_);
+                beta_2.push(a.a);
+            }
+            // first shuffle
+            let mut shuffled_1 = Vec::with_capacity(len);
+            for (pi, alpha) in pi.iter().zip(alpha_1) {
+                let pi_1 = pi.b.0 as usize;
+                shuffled_1.push(beta_2[pi_1] + alpha);
+            }
+            io_context.network.send_next_many(&shuffled_1)?;
+            let delta = io_context.network.recv_prev_many()?;
```

Comment:
can be one call 
https://github.com/TaceoLabs/co-snarks/blob/c04d2c0d3172aa790f18f4776ba36b248261565e/mpc-core/src/protocols/rep3/network.rs#L164

Reply:
Generelly if we have `send_next` and `recv_prev` we can do it with this

Reply:
replaced